### DISCUSSION TITLE
Legacy BIOS support

### DISCUSSION THREAD

**Discussion by oskarirauta** [RefindPlusRepo/RefindPlusUDK#13] [General]
Hi.

So far, I haven't compiled this, since I just found it- But soon I have. And then I have chance to try if this works better than other solutions.

What I then am looking for? Well, I'd like to use a pretty graphical environment as a boot manager, refind/clover/... checks out on this. But I'd like to have same manager for both, uefi AND legacy Bios. Clover attempts to succeed in that. But clover is out of the question, as I am looking for maximum compatibility, and Clover does not work at all even in UEFI boot method in BHyve virtual environment. Refind does not have this issue (not sure yet, if RefindPlus has- finding out soon enough..)

All UEFI emulation solutions (tianocore edk/udk) seem to fail more or less when tested on Apple machine running VMWare Fusion, except one, that works only partially and reason is known why partially. I have to admit, I am not expert on tianocore, nor have I used assembler more than few testing attempts about 25 years ago. Anyway, none of solutions available that I have found, are ended up in a complete failure to boot on vmware fusion machine using legacy bios, so far - and yes, I have tried with both tianocore emulator and clover.

Only working solution is available from Clover and from ONE specific version contains binary that boots, but unfortunately input devices cannot be used. Reason is pretty self explanatory also, it seems that when I check the sources, ISA support (including ps/2 support) does not seem to be enabled when this build has been created- so when booting virtual machine in Legacy BIOS mode- it kinda makes sense, that input devices also use legacy methods- such as ps/2 - and this is not configurable feature, and ofcourse, it would be better anyway if it would had been enabled/supported.

Well, anyway, that's the only solution that I have been able to find which boots and starts PE32 executable, in this case cloverx64.efi, ofcourse I'd like to change the source and make it try to boot biosx64.efi, and if not available, bootx64.efi, so I could use that environment to boot whatever- and in my case, Refind (or RefindPlus)..

It is available from Clover's legacy site on SourceForge, on this page: [Bootable ISO's on SourceForge.net](https://sourceforge.net/projects/cloverefiboot/files/Bootable_ISO/)

and it's build r4769. All past builds, or newer builds seem to fail to boot (classic 6 on screen and halt there). That build shows clover even with animation, unfortunately you cannot boot anything as none of input devices is available due to missing ps/2 support.

I have attempted to build r4769 by myself- but used tianocore revision is unknown, and most likely, build environment might also have changed things around. I built it on ArchLinux, searched for newest tianocore revision on date when r4769, succesfully was able to build, got identically sized boot6 file and all, but not matching sha256 sum and with failed boot, even when built "as-is", which means with "same options", including missing ps/2 support- so, no go there..

I do not have very high hopes on success with RefindPlusUDK either, but I am willing to try (installing podman as we speak on my Mac to use docker environment to build it, slow connection, takes some time to get there..)

UDK is pretty much (unfortunately) described as dead, unless we speak about QEmu.

Why this then is important. Well, if you get yourself a (vps) server, they are running in a virtual environments, some are docker kind of solutions, but better ones are not. Unfortunately they run with Legacy BIOS on 99% of times. And I'd really would love to support both, legacy bios AND UEFI. And use same environment for both..

There actually is a solution for this, which I've used in past, but I'd love to have some newer stuff, where Refind would come to the picture. I used to use SysLinux, which is available for both, uefi and legacy bios, and can use same configuration. But I'd rather got with refind.

Wish me luck....

**Comment by dakanji** [comment]
> Wish me luck....

Terve ... Good luck!

**Reply by oskarirauta** [reply]
Terve, ja hyvät joulut ;)

**Reply by oskarirauta** [reply]
Means hello and merry christmas for those who don't speak finnish..
